In [58]:
# Import libraries

import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

# pd.set_option('display.max_rows',None)

## Importing dataset

In [59]:
## company_order_report

In [60]:
company_order_report = pd.read_excel("Company X - Order Report.xlsx")
company_order_report

ExternOrderNo            SKU  Order Qty
0       2001827036  8904223818706        1.0
1       2001827036  8904223819093        1.0
2       2001827036  8904223819109        1.0
3       2001827036  8904223818430        1.0
4       2001827036  8904223819277        1.0
..             ...            ...        ...
395     2001806229  8904223818942        1.0
396     2001806229  8904223818850        1.0
397     2001806226  8904223818850        2.0
398     2001806210  8904223816214        1.0
399     2001806210  8904223818874        1.0

[400 rows x 3 columns]

In [61]:
company_invoice = pd.read_excel("Courier Company - Invoice.xlsx")
company_invoice

AWB Code    Order ID  Charged Weight  Warehouse Pincode  \
0    1091117222124  2001806232            1.30             121003   
1    1091117222194  2001806273            1.00             121003   
2    1091117222931  2001806408            2.50             121003   
3    1091117223244  2001806458            1.00             121003   
4    1091117229345  2001807012            0.15             121003   
..             ...         ...             ...                ...   
119  1091118551656  2001812941            0.73             121003   
120  1091117614452  2001809383            0.50             121003   
121  1091120922803  2001820978            0.50             121003   
122  1091121844806  2001811475            0.50             121003   
123  1091121846136  2001811305            0.50             121003   

     Customer Pincode Zone         Type of Shipment  Billing Amount (Rs.)  
0              507101    d          Forward charges                 135.0  
1              486886    d          Forward charges                  90.2  
2              532484    d          Forward charges                 224.6  
3              143001    b          Forward charges                  61.3  
4              515591    d          Forward charges                  45.4  
..                ...  ...                      ...                   ...  
119            325207    d          Forward charges                  90.2  
120            303702    d  Forward and RTO charges                  86.7  
121            313301    d          Forward charges                  45.4  
122            173212    b          Forward charges                  33.0  
123            302020    d          Forward charges                  45.4  

[124 rows x 8 columns]

In [62]:
company_master = pd.read_excel("Company X - SKU Master.xlsx")
company_master

SKU  Weight (g)
0   8904223815682         210
1   8904223815859         165
2   8904223815866         113
3   8904223815873          65
4   8904223816214         120
..            ...         ...
61  8904223819505         210
62  8904223819499         210
63  8904223819512         210
64  8904223819543         300
65     SACHETS001          10

[66 rows x 2 columns]

In [64]:
#courier_company_rates
courier_company_rates = pd.read_excel("Courier Company - Rates.xlsx")
courier_company_rates

Zone  Weight Slabs  Forward Fixed Charge  \
0    A          0.50                  29.5   
1    B          1.00                  33.0   
2    C          1.25                  40.1   
3    D          1.50                  45.4   
4    E          2.00                  56.6   

   Forward Additional Weight Slab Charge  RTO Fixed Charge  \
0                                   23.6              13.6   
1                                   28.3              20.5   
2                                   38.9              31.9   
3                                   44.8              41.3   
4                                   55.5              50.7   

   RTO Additional Weight Slab Charge  
0                               23.6  
1                               28.3  
2                               38.9  
3                               44.8  
4                               55.5

In [65]:
# company_zones
company_zones = pd.read_excel("Company X - Pincode Zones.xlsx")
company_zones

Warehouse Pincode  Customer Pincode Zone
0               121003            507101    d
1               121003            486886    d
2               121003            532484    d
3               121003            143001    b
4               121003            515591    d
..                 ...               ...  ...
119             121003            325207    b
120             121003            303702    b
121             121003            313301    b
122             121003            173212    e
123             121003            302020    b

[124 rows x 3 columns]

In [66]:
# renaming the column zone as zone_x
company_zones.rename(columns = {'Zone':'zone_X'}, inplace = True)
company_zones

Warehouse Pincode  Customer Pincode zone_X
0               121003            507101      d
1               121003            486886      d
2               121003            532484      d
3               121003            143001      b
4               121003            515591      d
..                 ...               ...    ...
119             121003            325207      b
120             121003            303702      b
121             121003            313301      b
122             121003            173212      e
123             121003            302020      b

[124 rows x 3 columns]

In [67]:
#  Merging SKU_Master and Order_report Datasets

In [68]:
sku_weight = pd.merge(company_master , company_order_report, on = 'SKU',how="outer" )
sku_weight

SKU  Weight (g)  ExternOrderNo  Order Qty
0    8904223815682         210     2001808801        1.0
1    8904223815859         165     2001820978        1.0
2    8904223815859         165     2001811363        1.0
3    8904223815859         165     2001811363        1.0
4    8904223815859         165     2001808801        1.0
..             ...         ...            ...        ...
396  8904223819512         210     2001806733        1.0
397  8904223819543         300     2001811306        1.0
398     SACHETS001          10     2001827036        1.0
399     SACHETS001          10     2001825261        1.0
400     SACHETS001          10     2001819252        1.0

[401 rows x 4 columns]

In [69]:
# Calulating the total weight of the order by its total quantity

In [70]:
sku_weight['weight_by_quantity'] = sku_weight['Weight (g)'] * sku_weight['Order Qty']
sku_weight

SKU  Weight (g)  ExternOrderNo  Order Qty  weight_by_quantity
0    8904223815682         210     2001808801        1.0               210.0
1    8904223815859         165     2001820978        1.0               165.0
2    8904223815859         165     2001811363        1.0               165.0
3    8904223815859         165     2001811363        1.0               165.0
4    8904223815859         165     2001808801        1.0               165.0
..             ...         ...            ...        ...                 ...
396  8904223819512         210     2001806733        1.0               210.0
397  8904223819543         300     2001811306        1.0               300.0
398     SACHETS001          10     2001827036        1.0                10.0
399     SACHETS001          10     2001825261        1.0                10.0
400     SACHETS001          10     2001819252        1.0                10.0

[401 rows x 5 columns]

### Total weight as per x

In [72]:
sku_weight

SKU  Weight (g)  ExternOrderNo  Order Qty  weight_by_quantity
0    8904223815682         210     2001808801        1.0               210.0
1    8904223815859         165     2001820978        1.0               165.0
2    8904223815859         165     2001811363        1.0               165.0
3    8904223815859         165     2001811363        1.0               165.0
4    8904223815859         165     2001808801        1.0               165.0
..             ...         ...            ...        ...                 ...
396  8904223819512         210     2001806733        1.0               210.0
397  8904223819543         300     2001811306        1.0               300.0
398     SACHETS001          10     2001827036        1.0                10.0
399     SACHETS001          10     2001825261        1.0                10.0
400     SACHETS001          10     2001819252        1.0                10.0

[401 rows x 5 columns]

In [73]:
tmp = sku_weight.groupby('ExternOrderNo')['weight_by_quantity'].sum()
tmp = tmp.reset_index()

In [74]:
# Renaming the weigth_by_quantity column

In [75]:
tmp.rename(columns = {'weight_by_quantity':'total_wt_X'}, inplace = True)
tmp

ExternOrderNo  total_wt_X
0       2001806210       220.0
1       2001806226       480.0
2       2001806229       500.0
3       2001806232      1302.0
4       2001806233       245.0
..             ...         ...
119     2001821995       477.0
120     2001822466      1376.0
121     2001823564       672.0
122     2001825261      1557.0
123     2001827036      2176.0

[124 rows x 2 columns]

### weight slab as per x

In [76]:
tmp['d_500'] = tmp['total_wt_X']/500
tmp['d_500'] = np.ceil(tmp['d_500'])
tmp['weight_slab_X'] = tmp['d_500'] * 0.5
tmp

ExternOrderNo  total_wt_X  d_500  weight_slab_X
0       2001806210       220.0    1.0            0.5
1       2001806226       480.0    1.0            0.5
2       2001806229       500.0    1.0            0.5
3       2001806232      1302.0    3.0            1.5
4       2001806233       245.0    1.0            0.5
..             ...         ...    ...            ...
119     2001821995       477.0    1.0            0.5
120     2001822466      1376.0    3.0            1.5
121     2001823564       672.0    2.0            1.0
122     2001825261      1557.0    4.0            2.0
123     2001827036      2176.0    5.0            2.5

[124 rows x 4 columns]

In [77]:
## Renaming the ExternalOrderNo to Order_ID

In [78]:
tmp.rename(columns = {'ExternOrderNo':'Order ID'}, inplace = True)
tmp

Order ID  total_wt_X  d_500  weight_slab_X
0    2001806210       220.0    1.0            0.5
1    2001806226       480.0    1.0            0.5
2    2001806229       500.0    1.0            0.5
3    2001806232      1302.0    3.0            1.5
4    2001806233       245.0    1.0            0.5
..          ...         ...    ...            ...
119  2001821995       477.0    1.0            0.5
120  2001822466      1376.0    3.0            1.5
121  2001823564       672.0    2.0            1.0
122  2001825261      1557.0    4.0            2.0
123  2001827036      2176.0    5.0            2.5

[124 rows x 4 columns]

In [79]:
tmp_file = pd.merge(tmp,company_invoice , on ='Order ID' , how = 'outer' )
tmp_file

Order ID  total_wt_X  d_500  weight_slab_X       AWB Code  \
0    2001806210       220.0    1.0            0.5  1091117221940   
1    2001806226       480.0    1.0            0.5  1091117222065   
2    2001806229       500.0    1.0            0.5  1091117222080   
3    2001806232      1302.0    3.0            1.5  1091117222124   
4    2001806233       245.0    1.0            0.5  1091117222135   
..          ...         ...    ...            ...            ...   
119  2001821995       477.0    1.0            0.5  1091121183730   
120  2001822466      1376.0    3.0            1.5  1091121305541   
121  2001823564       672.0    2.0            1.0  1091121666133   
122  2001825261      1557.0    4.0            2.0  1091121981575   
123  2001827036      2176.0    5.0            2.5  1091122418320   

     Charged Weight  Warehouse Pincode  Customer Pincode Zone  \
0              2.92             121003            140604    b   
1              0.68             121003            723146    d   
2              0.71             121003            421204    d   
3              1.30             121003            507101    d   
4              0.78             121003            263139    b   
..              ...                ...               ...  ...   
119            0.50             121003            342008    d   
120            1.10             121003            342301    d   
121            0.70             121003            492001    d   
122            1.60             121003            517128    d   
123            1.60             121003            173213    b   

            Type of Shipment  Billing Amount (Rs.)  
0            Forward charges                 174.5  
1            Forward charges                  90.2  
2            Forward charges                  90.2  
3            Forward charges                 135.0  
4            Forward charges                  61.3  
..                       ...                   ...  
119          Forward charges                  45.4  
120          Forward charges                 135.0  
121  Forward and RTO charges                 172.8  
122  Forward and RTO charges                 345.0  
123          Forward charges                 117.9  

[124 rows x 11 columns]

In [80]:
## Renaming multiple columns

In [81]:
tmp_file.rename(columns = {'Charged Weight':'total_wt_cc'}, inplace = True)

tmp_file.rename(columns = {'Zone':'zone_cc'}, inplace = True)

tmp_file.rename(columns = {'Billing Amount (Rs.)':'charges_by_cc'}, inplace = True)
tmp_file

Order ID  total_wt_X  d_500  weight_slab_X       AWB Code  total_wt_cc  \
0    2001806210       220.0    1.0            0.5  1091117221940         2.92   
1    2001806226       480.0    1.0            0.5  1091117222065         0.68   
2    2001806229       500.0    1.0            0.5  1091117222080         0.71   
3    2001806232      1302.0    3.0            1.5  1091117222124         1.30   
4    2001806233       245.0    1.0            0.5  1091117222135         0.78   
..          ...         ...    ...            ...            ...          ...   
119  2001821995       477.0    1.0            0.5  1091121183730         0.50   
120  2001822466      1376.0    3.0            1.5  1091121305541         1.10   
121  2001823564       672.0    2.0            1.0  1091121666133         0.70   
122  2001825261      1557.0    4.0            2.0  1091121981575         1.60   
123  2001827036      2176.0    5.0            2.5  1091122418320         1.60   

     Warehouse Pincode  Customer Pincode zone_cc         Type of Shipment  \
0               121003            140604       b          Forward charges   
1               121003            723146       d          Forward charges   
2               121003            421204       d          Forward charges   
3               121003            507101       d          Forward charges   
4               121003            263139       b          Forward charges   
..                 ...               ...     ...                      ...   
119             121003            342008       d          Forward charges   
120             121003            342301       d          Forward charges   
121             121003            492001       d  Forward and RTO charges   
122             121003            517128       d  Forward and RTO charges   
123             121003            173213       b          Forward charges   

     charges_by_cc  
0            174.5  
1             90.2  
2             90.2  
3            135.0  
4             61.3  
..             ...  
119           45.4  
120          135.0  
121          172.8  
122          345.0  
123          117.9  

[124 rows x 11 columns]

In [83]:
## Dropping unwanted columns

In [84]:
tmp_file.drop(["d_500"], axis =1,inplace = True)
tmp_file.drop(["Warehouse Pincode"], axis =1,inplace = True)

In [85]:
tmp_file

Order ID  total_wt_X  weight_slab_X       AWB Code  total_wt_cc  \
0    2001806210       220.0            0.5  1091117221940         2.92   
1    2001806226       480.0            0.5  1091117222065         0.68   
2    2001806229       500.0            0.5  1091117222080         0.71   
3    2001806232      1302.0            1.5  1091117222124         1.30   
4    2001806233       245.0            0.5  1091117222135         0.78   
..          ...         ...            ...            ...          ...   
119  2001821995       477.0            0.5  1091121183730         0.50   
120  2001822466      1376.0            1.5  1091121305541         1.10   
121  2001823564       672.0            1.0  1091121666133         0.70   
122  2001825261      1557.0            2.0  1091121981575         1.60   
123  2001827036      2176.0            2.5  1091122418320         1.60   

     Customer Pincode zone_cc         Type of Shipment  charges_by_cc  
0              140604       b          Forward charges          174.5  
1              723146       d          Forward charges           90.2  
2              421204       d          Forward charges           90.2  
3              507101       d          Forward charges          135.0  
4              263139       b          Forward charges           61.3  
..                ...     ...                      ...            ...  
119            342008       d          Forward charges           45.4  
120            342301       d          Forward charges          135.0  
121            492001       d  Forward and RTO charges          172.8  
122            517128       d  Forward and RTO charges          345.0  
123            173213       b          Forward charges          117.9  

[124 rows x 9 columns]

In [86]:
## Converting the total weight from cubic centimeter to KG
tmp_file['total_wt_cc'] = tmp_file['total_wt_cc'].apply(lambda x:x*1000)
tmp_file

Order ID  total_wt_X  weight_slab_X       AWB Code  total_wt_cc  \
0    2001806210       220.0            0.5  1091117221940       2920.0   
1    2001806226       480.0            0.5  1091117222065        680.0   
2    2001806229       500.0            0.5  1091117222080        710.0   
3    2001806232      1302.0            1.5  1091117222124       1300.0   
4    2001806233       245.0            0.5  1091117222135        780.0   
..          ...         ...            ...            ...          ...   
119  2001821995       477.0            0.5  1091121183730        500.0   
120  2001822466      1376.0            1.5  1091121305541       1100.0   
121  2001823564       672.0            1.0  1091121666133        700.0   
122  2001825261      1557.0            2.0  1091121981575       1600.0   
123  2001827036      2176.0            2.5  1091122418320       1600.0   

     Customer Pincode zone_cc         Type of Shipment  charges_by_cc  
0              140604       b          Forward charges          174.5  
1              723146       d          Forward charges           90.2  
2              421204       d          Forward charges           90.2  
3              507101       d          Forward charges          135.0  
4              263139       b          Forward charges           61.3  
..                ...     ...                      ...            ...  
119            342008       d          Forward charges           45.4  
120            342301       d          Forward charges          135.0  
121            492001       d  Forward and RTO charges          172.8  
122            517128       d  Forward and RTO charges          345.0  
123            173213       b          Forward charges          117.9  

[124 rows x 9 columns]

### weight slab as per cc

In [87]:
tmp_file['d_500'] = tmp_file['total_wt_cc']/500
tmp_file['d_500'] = np.ceil(tmp_file['d_500'])
tmp_file['weight_slab_cc'] = tmp_file['d_500'] * 0.5
tmp_file

Order ID  total_wt_X  weight_slab_X       AWB Code  total_wt_cc  \
0    2001806210       220.0            0.5  1091117221940       2920.0   
1    2001806226       480.0            0.5  1091117222065        680.0   
2    2001806229       500.0            0.5  1091117222080        710.0   
3    2001806232      1302.0            1.5  1091117222124       1300.0   
4    2001806233       245.0            0.5  1091117222135        780.0   
..          ...         ...            ...            ...          ...   
119  2001821995       477.0            0.5  1091121183730        500.0   
120  2001822466      1376.0            1.5  1091121305541       1100.0   
121  2001823564       672.0            1.0  1091121666133        700.0   
122  2001825261      1557.0            2.0  1091121981575       1600.0   
123  2001827036      2176.0            2.5  1091122418320       1600.0   

     Customer Pincode zone_cc         Type of Shipment  charges_by_cc  d_500  \
0              140604       b          Forward charges          174.5    6.0   
1              723146       d          Forward charges           90.2    2.0   
2              421204       d          Forward charges           90.2    2.0   
3              507101       d          Forward charges          135.0    3.0   
4              263139       b          Forward charges           61.3    2.0   
..                ...     ...                      ...            ...    ...   
119            342008       d          Forward charges           45.4    1.0   
120            342301       d          Forward charges          135.0    3.0   
121            492001       d  Forward and RTO charges          172.8    2.0   
122            517128       d  Forward and RTO charges          345.0    4.0   
123            173213       b          Forward charges          117.9    4.0   

     weight_slab_cc  
0               3.0  
1               1.0  
2               1.0  
3               1.5  
4               1.0  
..              ...  
119             0.5  
120             1.5  
121             1.0  
122             2.0  
123             2.0  

[124 rows x 11 columns]

In [88]:
## Dropping a column
tmp_file.drop(["d_500"], axis =1,inplace = True)
tmp_file

Order ID  total_wt_X  weight_slab_X       AWB Code  total_wt_cc  \
0    2001806210       220.0            0.5  1091117221940       2920.0   
1    2001806226       480.0            0.5  1091117222065        680.0   
2    2001806229       500.0            0.5  1091117222080        710.0   
3    2001806232      1302.0            1.5  1091117222124       1300.0   
4    2001806233       245.0            0.5  1091117222135        780.0   
..          ...         ...            ...            ...          ...   
119  2001821995       477.0            0.5  1091121183730        500.0   
120  2001822466      1376.0            1.5  1091121305541       1100.0   
121  2001823564       672.0            1.0  1091121666133        700.0   
122  2001825261      1557.0            2.0  1091121981575       1600.0   
123  2001827036      2176.0            2.5  1091122418320       1600.0   

     Customer Pincode zone_cc         Type of Shipment  charges_by_cc  \
0              140604       b          Forward charges          174.5   
1              723146       d          Forward charges           90.2   
2              421204       d          Forward charges           90.2   
3              507101       d          Forward charges          135.0   
4              263139       b          Forward charges           61.3   
..                ...     ...                      ...            ...   
119            342008       d          Forward charges           45.4   
120            342301       d          Forward charges          135.0   
121            492001       d  Forward and RTO charges          172.8   
122            517128       d  Forward and RTO charges          345.0   
123            173213       b          Forward charges          117.9   

     weight_slab_cc  
0               3.0  
1               1.0  
2               1.0  
3               1.5  
4               1.0  
..              ...  
119             0.5  
120             1.5  
121             1.0  
122             2.0  
123             2.0  

[124 rows x 10 columns]

In [89]:
tmp_file2 = pd.merge(tmp_file , company_zones , on = 'Customer Pincode',how="left" )
tmp_file2

Order ID  total_wt_X  weight_slab_X       AWB Code  total_wt_cc  \
0    2001806210       220.0            0.5  1091117221940       2920.0   
1    2001806226       480.0            0.5  1091117222065        680.0   
2    2001806229       500.0            0.5  1091117222080        710.0   
3    2001806232      1302.0            1.5  1091117222124       1300.0   
4    2001806233       245.0            0.5  1091117222135        780.0   
..          ...         ...            ...            ...          ...   
169  2001821995       477.0            0.5  1091121183730        500.0   
170  2001822466      1376.0            1.5  1091121305541       1100.0   
171  2001823564       672.0            1.0  1091121666133        700.0   
172  2001825261      1557.0            2.0  1091121981575       1600.0   
173  2001827036      2176.0            2.5  1091122418320       1600.0   

     Customer Pincode zone_cc         Type of Shipment  charges_by_cc  \
0              140604       b          Forward charges          174.5   
1              723146       d          Forward charges           90.2   
2              421204       d          Forward charges           90.2   
3              507101       d          Forward charges          135.0   
4              263139       b          Forward charges           61.3   
..                ...     ...                      ...            ...   
169            342008       d          Forward charges           45.4   
170            342301       d          Forward charges          135.0   
171            492001       d  Forward and RTO charges          172.8   
172            517128       d  Forward and RTO charges          345.0   
173            173213       b          Forward charges          117.9   

     weight_slab_cc  Warehouse Pincode zone_X  
0               3.0             121003      b  
1               1.0             121003      d  
2               1.0             121003      d  
3               1.5             121003      d  
4               1.0             121003      b  
..              ...                ...    ...  
169             0.5             121003      b  
170             1.5             121003      b  
171             1.0             121003      d  
172             2.0             121003      d  
173             2.0             121003      e  

[174 rows x 12 columns]

In [90]:
## Dropping columns
tmp_file2.drop(["Warehouse Pincode"], axis =1,inplace = True)
tmp_file2.drop(["Customer Pincode"], axis =1,inplace = True)
tmp_file2

Order ID  total_wt_X  weight_slab_X       AWB Code  total_wt_cc  \
0    2001806210       220.0            0.5  1091117221940       2920.0   
1    2001806226       480.0            0.5  1091117222065        680.0   
2    2001806229       500.0            0.5  1091117222080        710.0   
3    2001806232      1302.0            1.5  1091117222124       1300.0   
4    2001806233       245.0            0.5  1091117222135        780.0   
..          ...         ...            ...            ...          ...   
169  2001821995       477.0            0.5  1091121183730        500.0   
170  2001822466      1376.0            1.5  1091121305541       1100.0   
171  2001823564       672.0            1.0  1091121666133        700.0   
172  2001825261      1557.0            2.0  1091121981575       1600.0   
173  2001827036      2176.0            2.5  1091122418320       1600.0   

    zone_cc         Type of Shipment  charges_by_cc  weight_slab_cc zone_X  
0         b          Forward charges          174.5             3.0      b  
1         d          Forward charges           90.2             1.0      d  
2         d          Forward charges           90.2             1.0      d  
3         d          Forward charges          135.0             1.5      d  
4         b          Forward charges           61.3             1.0      b  
..      ...                      ...            ...             ...    ...  
169       d          Forward charges           45.4             0.5      b  
170       d          Forward charges          135.0             1.5      b  
171       d  Forward and RTO charges          172.8             1.0      d  
172       d  Forward and RTO charges          345.0             2.0      d  
173       b          Forward charges          117.9             2.0      e  

[174 rows x 10 columns]

In [91]:
tmp_file2.isnull().sum()

Order ID            0
total_wt_X          0
weight_slab_X       0
AWB Code            0
total_wt_cc         0
zone_cc             0
Type of Shipment    0
charges_by_cc       0
weight_slab_cc      0
zone_X              0
dtype: int64

In [92]:
tmp_file2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 174 entries, 0 to 173
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Order ID          174 non-null    int64  
 1   total_wt_X        174 non-null    float64
 2   weight_slab_X     174 non-null    float64
 3   AWB Code          174 non-null    int64  
 4   total_wt_cc       174 non-null    float64
 5   zone_cc           174 non-null    object 
 6   Type of Shipment  174 non-null    object 
 7   charges_by_cc     174 non-null    float64
 8   weight_slab_cc    174 non-null    float64
 9   zone_X            174 non-null    object 
dtypes: float64(5), int64(2), object(3)
memory usage: 13.7+ KB


In [94]:
tmp_file2['total_wt_X'] = tmp_file2['total_wt_X'].astype('int64')

In [95]:
tmp_file2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 174 entries, 0 to 173
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Order ID          174 non-null    int64  
 1   total_wt_X        174 non-null    int64  
 2   weight_slab_X     174 non-null    float64
 3   AWB Code          174 non-null    int64  
 4   total_wt_cc       174 non-null    float64
 5   zone_cc           174 non-null    object 
 6   Type of Shipment  174 non-null    object 
 7   charges_by_cc     174 non-null    float64
 8   weight_slab_cc    174 non-null    float64
 9   zone_X            174 non-null    object 
dtypes: float64(4), int64(3), object(3)
memory usage: 13.7+ KB


In [96]:
list_zones = tmp_file2['zone_X'].unique()
list_zones

array(['b', 'd', 'e'], dtype=object)

In [97]:
final_df = pd.DataFrame()
final_df

Empty DataFrame
Columns: []
Index: []

In [98]:
value_pair = {'a':29.5,'b':33,'c':40.1,'d':45.4,'e':56.6}
additional_charges = {'a':23.6,'b':28.3,'c':38.9,'d':44.8,'e':55.5}
rto_fixed_charges = {'a':13.6,'b':20.5,'c':31.9,'d':41.3,'e':50.7}
rto_additional_charges = {'a':23.6,'b':28.3,'c':38.9,'d':44.8,'e':55.5}

In [99]:
for i in list_zones:
    temp3 = tmp_file2[tmp_file2['zone_X'] == i]
    temp3['fixed'] = value_pair[i]
    temp3['additional'] = np.where(temp3 ['weight_slab_X']*2 > 1,((temp3['weight_slab_X']*2)-1)*additional_charges[i],0)
    temp3['rto_fixed_charges'] =  np.where(temp3['Type of Shipment'].str.contains("RTO", case=False), rto_fixed_charges[i], 0)
    temp3['rto_additional_charges'] =  np.where((temp3 ['weight_slab_X']*2 > 1) & (temp3['Type of Shipment'].str.contains('Forward and RTO charges', case=False)),((temp3['weight_slab_X']*2)-1)*rto_additional_charges[i],0)
    final_df = pd.concat([final_df,temp3])

In [100]:
final_df.reset_index(inplace = True,drop = True)
final_df

Order ID  total_wt_X  weight_slab_X       AWB Code  total_wt_cc  \
0    2001806210         220            0.5  1091117221940       2920.0   
1    2001806233         245            0.5  1091117222135        780.0   
2    2001806304         500            0.5  1091117222360        710.0   
3    2001806304         500            0.5  1091117222360        710.0   
4    2001806304         500            0.5  1091117222360        710.0   
..          ...         ...            ...            ...          ...   
169  2001811604         721            1.0  1091118004245        800.0   
170  2001811604         721            1.0  1091118004245        800.0   
171  2001818390         841            1.0  1091120014461        800.0   
172  2001819252         270            0.5  1091120352712        300.0   
173  2001827036        2176            2.5  1091122418320       1600.0   

    zone_cc         Type of Shipment  charges_by_cc  weight_slab_cc zone_X  \
0         b          Forward charges          174.5             3.0      b   
1         b          Forward charges           61.3             1.0      b   
2         d          Forward charges           90.2             1.0      b   
3         d          Forward charges           90.2             1.0      b   
4         d          Forward charges           90.2             1.0      b   
..      ...                      ...            ...             ...    ...   
169       b          Forward charges           61.3             1.0      e   
170       b          Forward charges           61.3             1.0      e   
171       e  Forward and RTO charges          213.5             1.0      e   
172       b          Forward charges           33.0             0.5      e   
173       b          Forward charges          117.9             2.0      e   

     fixed  additional  rto_fixed_charges  rto_additional_charges  
0     33.0         0.0                0.0                     0.0  
1     33.0         0.0                0.0                     0.0  
2     33.0         0.0                0.0                     0.0  
3     33.0         0.0                0.0                     0.0  
4     33.0         0.0                0.0                     0.0  
..     ...         ...                ...                     ...  
169   56.6        55.5                0.0                     0.0  
170   56.6        55.5                0.0                     0.0  
171   56.6        55.5               50.7                    55.5  
172   56.6         0.0                0.0                     0.0  
173   56.6       222.0                0.0                     0.0  

[174 rows x 14 columns]

In [101]:
#  Calculating the final expected charges
final_df['expected_charges_X'] = final_df['fixed'] + final_df['additional'] + final_df['rto_fixed_charges'] + final_df['rto_additional_charges']
final_df

Order ID  total_wt_X  weight_slab_X       AWB Code  total_wt_cc  \
0    2001806210         220            0.5  1091117221940       2920.0   
1    2001806233         245            0.5  1091117222135        780.0   
2    2001806304         500            0.5  1091117222360        710.0   
3    2001806304         500            0.5  1091117222360        710.0   
4    2001806304         500            0.5  1091117222360        710.0   
..          ...         ...            ...            ...          ...   
169  2001811604         721            1.0  1091118004245        800.0   
170  2001811604         721            1.0  1091118004245        800.0   
171  2001818390         841            1.0  1091120014461        800.0   
172  2001819252         270            0.5  1091120352712        300.0   
173  2001827036        2176            2.5  1091122418320       1600.0   

    zone_cc         Type of Shipment  charges_by_cc  weight_slab_cc zone_X  \
0         b          Forward charges          174.5             3.0      b   
1         b          Forward charges           61.3             1.0      b   
2         d          Forward charges           90.2             1.0      b   
3         d          Forward charges           90.2             1.0      b   
4         d          Forward charges           90.2             1.0      b   
..      ...                      ...            ...             ...    ...   
169       b          Forward charges           61.3             1.0      e   
170       b          Forward charges           61.3             1.0      e   
171       e  Forward and RTO charges          213.5             1.0      e   
172       b          Forward charges           33.0             0.5      e   
173       b          Forward charges          117.9             2.0      e   

     fixed  additional  rto_fixed_charges  rto_additional_charges  \
0     33.0         0.0                0.0                     0.0   
1     33.0         0.0                0.0                     0.0   
2     33.0         0.0                0.0                     0.0   
3     33.0         0.0                0.0                     0.0   
4     33.0         0.0                0.0                     0.0   
..     ...         ...                ...                     ...   
169   56.6        55.5                0.0                     0.0   
170   56.6        55.5                0.0                     0.0   
171   56.6        55.5               50.7                    55.5   
172   56.6         0.0                0.0                     0.0   
173   56.6       222.0                0.0                     0.0   

     expected_charges_X  
0                  33.0  
1                  33.0  
2                  33.0  
3                  33.0  
4                  33.0  
..                  ...  
169               112.1  
170               112.1  
171               218.3  
172                56.6  
173               278.6  

[174 rows x 15 columns]

In [102]:
#  Calculating the difference between expected charges and actual chages
final_df['difference_x_cc'] = round(final_df['expected_charges_X'] - final_df['charges_by_cc'],2)
final_df

Order ID  total_wt_X  weight_slab_X       AWB Code  total_wt_cc  \
0    2001806210         220            0.5  1091117221940       2920.0   
1    2001806233         245            0.5  1091117222135        780.0   
2    2001806304         500            0.5  1091117222360        710.0   
3    2001806304         500            0.5  1091117222360        710.0   
4    2001806304         500            0.5  1091117222360        710.0   
..          ...         ...            ...            ...          ...   
169  2001811604         721            1.0  1091118004245        800.0   
170  2001811604         721            1.0  1091118004245        800.0   
171  2001818390         841            1.0  1091120014461        800.0   
172  2001819252         270            0.5  1091120352712        300.0   
173  2001827036        2176            2.5  1091122418320       1600.0   

    zone_cc         Type of Shipment  charges_by_cc  weight_slab_cc zone_X  \
0         b          Forward charges          174.5             3.0      b   
1         b          Forward charges           61.3             1.0      b   
2         d          Forward charges           90.2             1.0      b   
3         d          Forward charges           90.2             1.0      b   
4         d          Forward charges           90.2             1.0      b   
..      ...                      ...            ...             ...    ...   
169       b          Forward charges           61.3             1.0      e   
170       b          Forward charges           61.3             1.0      e   
171       e  Forward and RTO charges          213.5             1.0      e   
172       b          Forward charges           33.0             0.5      e   
173       b          Forward charges          117.9             2.0      e   

     fixed  additional  rto_fixed_charges  rto_additional_charges  \
0     33.0         0.0                0.0                     0.0   
1     33.0         0.0                0.0                     0.0   
2     33.0         0.0                0.0                     0.0   
3     33.0         0.0                0.0                     0.0   
4     33.0         0.0                0.0                     0.0   
..     ...         ...                ...                     ...   
169   56.6        55.5                0.0                     0.0   
170   56.6        55.5                0.0                     0.0   
171   56.6        55.5               50.7                    55.5   
172   56.6         0.0                0.0                     0.0   
173   56.6       222.0                0.0                     0.0   

     expected_charges_X  difference_x_cc  
0                  33.0           -141.5  
1                  33.0            -28.3  
2                  33.0            -57.2  
3                  33.0            -57.2  
4                  33.0            -57.2  
..                  ...              ...  
169               112.1             50.8  
170               112.1             50.8  
171               218.3              4.8  
172                56.6             23.6  
173               278.6            160.7  

[174 rows x 16 columns]

#### Droping down the unwanted column in the dataset

In [103]:
final_df = final_df.drop(columns = ['fixed','additional','rto_fixed_charges','rto_additional_charges'])

In [104]:
df = final_df[['Order ID','AWB Code','total_wt_X','weight_slab_X','total_wt_cc','weight_slab_cc','zone_X','zone_cc','expected_charges_X','charges_by_cc','difference_x_cc']]
df

Order ID       AWB Code  total_wt_X  weight_slab_X  total_wt_cc  \
0    2001806210  1091117221940         220            0.5       2920.0   
1    2001806233  1091117222135         245            0.5        780.0   
2    2001806304  1091117222360         500            0.5        710.0   
3    2001806304  1091117222360         500            0.5        710.0   
4    2001806304  1091117222360         500            0.5        710.0   
..          ...            ...         ...            ...          ...   
169  2001811604  1091118004245         721            1.0        800.0   
170  2001811604  1091118004245         721            1.0        800.0   
171  2001818390  1091120014461         841            1.0        800.0   
172  2001819252  1091120352712         270            0.5        300.0   
173  2001827036  1091122418320        2176            2.5       1600.0   

     weight_slab_cc zone_X zone_cc  expected_charges_X  charges_by_cc  \
0               3.0      b       b                33.0          174.5   
1               1.0      b       b                33.0           61.3   
2               1.0      b       d                33.0           90.2   
3               1.0      b       d                33.0           90.2   
4               1.0      b       d                33.0           90.2   
..              ...    ...     ...                 ...            ...   
169             1.0      e       b               112.1           61.3   
170             1.0      e       b               112.1           61.3   
171             1.0      e       e               218.3          213.5   
172             0.5      e       b                56.6           33.0   
173             2.0      e       b               278.6          117.9   

     difference_x_cc  
0             -141.5  
1              -28.3  
2              -57.2  
3              -57.2  
4              -57.2  
..               ...  
169             50.8  
170             50.8  
171              4.8  
172             23.6  
173            160.7  

[174 rows x 11 columns]

### Find out the undercharges of greater than 0

In [105]:
under_charge= df.loc[df['difference_x_cc'] > 0]
under_charge

Order ID       AWB Code  total_wt_X  weight_slab_X  total_wt_cc  \
26   2001807415  1091117325094        1048            1.5       1000.0   
53   2001809383  1091117614452         607            1.0        500.0   
54   2001809592  1091117616121        1505            2.0       1500.0   
72   2001811305  1091121846136         750            1.0        500.0   
73   2001811305  1091121846136         750            1.0        500.0   
87   2001812838  1091118547832         558            1.0        600.0   
106  2001817093  1091121485824        1357            1.5       1300.0   
108  2001820978  1091120922803         515            1.0        500.0   
148  2001807362  1091117324394        2016            2.5       2000.0   
152  2001807976  1091117327496         721            1.0        700.0   
156  2001809794  1091117795531        1517            2.0       1500.0   
157  2001809820  1091117795623        3080            3.5       3000.0   
158  2001809917  1091121482593         630            1.0        600.0   
159  2001811192  1091117957780        1032            1.5       1130.0   
160  2001816684  1091119398844         920            1.0        990.0   
161  2001817160  1091119630264         700            1.0        700.0   
162  2001821190  1091120959015        1200            1.5       1200.0   
163  2001823564  1091121666133         672            1.0        700.0   
164  2001825261  1091121981575        1557            2.0       1600.0   
167  2001811475  1091121844806         689            1.0        500.0   
168  2001811475  1091121844806         689            1.0        500.0   
169  2001811604  1091118004245         721            1.0        800.0   
170  2001811604  1091118004245         721            1.0        800.0   
171  2001818390  1091120014461         841            1.0        800.0   
172  2001819252  1091120352712         270            0.5        300.0   
173  2001827036  1091122418320        2176            2.5       1600.0   

     weight_slab_cc zone_X zone_cc  expected_charges_X  charges_by_cc  \
26              1.0      b       b                89.6           61.3   
53              0.5      b       d               110.1           86.7   
54              1.5      b       b               117.9           89.6   
72              0.5      b       d                61.3           45.4   
73              0.5      b       d                61.3           45.4   
87              1.0      b       b               110.1          102.3   
106             1.5      b       b               166.7          151.1   
108             0.5      b       d                61.3           45.4   
148             2.0      d       d               224.6          179.8   
152             1.0      d       d               176.3          172.8   
156             1.5      d       d               179.8          135.0   
157             3.0      d       d               314.2          269.4   
158             1.0      d       d               176.3          172.8   
159             1.5      d       d               265.9          258.9   
160             1.0      d       d               176.3          172.8   
161             1.0      d       d               176.3          172.8   
162             1.5      d       d               265.9          258.9   
163             1.0      d       d               176.3          172.8   
164             2.0      d       d               355.5          345.0   
167             0.5      e       b               112.1           33.0   
168             0.5      e       b               112.1           33.0   
169             1.0      e       b               112.1           61.3   
170             1.0      e       b               112.1           61.3   
171             1.0      e       e               218.3          213.5   
172             0.5      e       b                56.6           33.0   
173             2.0      e       b               278.6          117.9   

     difference_x_cc  
26           

In [106]:
# Length of  overcharges
len(under_charge)

26

In [107]:
#  Find out the overchager of sum
under_charge.sum()

Order ID                             52047170665
AWB Code                          28369113007241
total_wt_X                                 27242
weight_slab_X                               36.0
total_wt_cc                              25820.0
weight_slab_cc                              30.0
zone_X                bbbbbbbbdddddddddddeeeeeee
zone_cc               bdbddbbddddddddddddbbbbebb
expected_charges_X                        4267.6
charges_by_cc                             3491.2
difference_x_cc                            776.4
dtype: object

1. `Analyzing the dataset we found out that the undercharges is 776.4 RS.`

In [108]:
## Dropping the column

In [109]:
over_charged = df.loc[df['difference_x_cc'] < 0]
over_charged

Order ID       AWB Code  total_wt_X  weight_slab_X  total_wt_cc  \
0    2001806210  1091117221940         220            0.5       2920.0   
1    2001806233  1091117222135         245            0.5        780.0   
2    2001806304  1091117222360         500            0.5        710.0   
3    2001806304  1091117222360         500            0.5        710.0   
4    2001806304  1091117222360         500            0.5        710.0   
..          ...            ...         ...            ...          ...   
141  2001806735  1091117226711         500            0.5        690.0   
142  2001807004  1091117229290         500            0.5        680.0   
144  2001807036  1091117323005        1459            1.5       1640.0   
145  2001807084  1091117323215         500            0.5        670.0   
166  2001808585  1091117436652         500            0.5        720.0   

     weight_slab_cc zone_X zone_cc  expected_charges_X  charges_by_cc  \
0               3.0      b       b                33.0          174.5   
1               1.0      b       b                33.0           61.3   
2               1.0      b       d                33.0           90.2   
3               1.0      b       d                33.0           90.2   
4               1.0      b       d                33.0           90.2   
..              ...    ...     ...                 ...            ...   
141             1.0      d       d                45.4           90.2   
142             1.0      d       d                45.4           90.2   
144             2.0      d       d               135.0          179.8   
145             1.0      d       d                45.4           90.2   
166             1.0      e       b                56.6           61.3   

     difference_x_cc  
0             -141.5  
1              -28.3  
2              -57.2  
3              -57.2  
4              -57.2  
..               ...  
141            -44.8  
142            -44.8  
144            -44.8  
145            -44.8  
166             -4.7  

[124 rows x 11 columns]

#### Find out the undercharge sum

In [110]:
over_charged.sum()

Order ID                                                   248224628165
AWB Code                                                135298668074743
total_wt_X                                                        76060
weight_slab_X                                                      97.5
total_wt_cc                                                    114550.0
weight_slab_cc                                                    148.5
zone_X                bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb...
zone_cc               bbddddddbbbbdbbbdddddddddddddddddddddddddddddd...
expected_charges_X                                               6405.5
charges_by_cc                                                   12992.1
difference_x_cc                                                 -6586.6
dtype: object

`analyzing the dataset overcharge iss 6586.6`

In [111]:
len(df.loc[df['difference_x_cc'] < 0])

124

In [112]:
fair_charged = df.loc[df['difference_x_cc'] == 0]
fair_charged

Order ID       AWB Code  total_wt_X  weight_slab_X  total_wt_cc  \
7    2001806458  1091117223244         700            1.0       1000.0   
19   2001806885  1091117229776         840            1.0       1000.0   
21   2001807058  1091117323112        1168            1.5       1150.0   
22   2001807058  1091117323112        1168            1.5       1150.0   
29   2001807814  1091117326612         607            1.0        790.0   
31   2001807956  1091117327275        1080            1.5       1080.0   
34   2001808118  1091117435134         343            0.5        500.0   
35   2001808118  1091117435134         343            0.5        500.0   
40   2001808507  1091117436383         607            1.0        790.0   
41   2001808675  1091117437050        1183            1.5       1200.0   
129  2001806232  1091117222124        1302            1.5       1300.0   
131  2001806273  1091117222194         615            1.0       1000.0   
133  2001806408  1091117222931        2265            2.5       2500.0   
138  2001806686  1091117229555         240            0.5        150.0   
143  2001807012  1091117229345         240            0.5        150.0   
146  2001807186  1091117323812         500            0.5        500.0   
147  2001807290  1091117324206         500            0.5        500.0   
149  2001807930  1091117327695         240            0.5        150.0   
150  2001807931  1091117327172         607            1.0        720.0   
151  2001807960  1091117327312         930            1.0       1000.0   
153  2001808102  1091117435005        1157            1.5       1280.0   
154  2001808207  1091117435370         607            1.0        790.0   
155  2001808542  1091117436464         734            1.0        860.0   
165  2001808295  1091117435661         245            0.5        200.0   

     weight_slab_cc zone_X zone_cc  expected_charges_X  charges_by_cc  \
7               1.0      b       b                61.3           61.3   
19              1.0      b       b                61.3           61.3   
21              1.5      b       b                89.6           89.6   
22              1.5      b       b                89.6           89.6   
29              1.0      b       b                61.3           61.3   
31              1.5      b       b                89.6           89.6   
34              0.5      b       b                33.0           33.0   
35              0.5      b       b                33.0           33.0   
40              1.0      b       b                61.3           61.3   
41              1.5      b       b                89.6           89.6   
129             1.5      d       d               135.0          135.0   
131             1.0      d       d                90.2           90.2   
133             2.5      d       d               224.6          224.6   
138             0.5      d       d                45.4           45.4   
143             0.5      d       d                45.4           45.4   
146             0.5      d       d                45.4           45.4   
147             0.5      d       d                45.4           45.4   
149             0.5      d       d                45.4           45.4   
150             1.0      d       d                90.2           90.2   
151             1.0      d       d                90.2           90.2   
153             1.5      d       d               135.0          135.0   
154             1.0      d       d                90.2           90.2   
155             1.0      d       d                90.2           90.2   
165             0.5      e       e               107.3          107.3   

     difference_x_cc  
7                0.0  
19               0.0  
21               0.0  
22               0.0  
29               0.0  
31               0.0  
34               0.0  
35               0.0  
40               0.0  
41               0.0  
129              0.0  
131             -0.0  
133              0.0  
138     

In [113]:
len(df.loc[df['difference_x_cc'] == 0])

24

### Find out the fair charge sum


In [114]:
fair_charged.sum()

Order ID                           48043380701
AWB Code                        26186815995678
total_wt_X                               18221
weight_slab_X                             24.5
total_wt_cc                            20260.0
weight_slab_cc                            24.5
zone_X                bbbbbbbbbbddddddddddddde
zone_cc               bbbbbbbbbbddddddddddddde
expected_charges_X                      1949.5
charges_by_cc                           1949.5
difference_x_cc                            0.0
dtype: object

`1. analyzing the dataset, fair charges is 0.`

In [115]:
## Converting the dataframe into the csv file
df.to_csv('df_file_csv', index=False)